### DISCUSSION TITLE
Returning Components from server actions for lazy loading - Valid or not?!!

### DISCUSSION THREAD

**Discussion by rohitpotato** [vercel/next.js#69751] [Help]
### Summary

Is returning server/client components from server-actions a correct approach? I have a usecase where I want to lazy load components via intersection observer, to do this, and reduce client bundles + processing, I came across [this](https://github.com/gabrielelpidio/next-infinite-scroll-server-actions) implementation.

We want to adopt this practice at scale, but is this the correct way of doing things?

Or, should we **dynamic import** components as it has been done traditionally?

What are the pros and cons? 

### Additional information

```js
https://github.com/gabrielelpidio/next-infinite-scroll-server-actions
```

**Comment by Juneezee** [comment]
~~Server Actions does not support returning JSX, see https://react.dev/reference/rsc/use-server#serializable-parameters-and-return-values~~

EDIT: See https://github.com/vercel/next.js/discussions/69751#discussioncomment-10564932 for updated answer

**Reply by rohitpotato** [reply]
@Juneezee But the above approach works? It sends over the components over the wire,  i see the rsc payload being sent over the network.

Am i missing something here?

**Reply by Juneezee** [reply]
@rohitpotato Sorry, my bad. I missed that one important sentence

> Supported serializable return values are the same as [serializable props](https://react.dev/reference/rsc/use-client#passing-props-from-server-to-client-components) for a boundary Client Component.


And according to https://react.dev/reference/rsc/use-client#serializable-types, 

> Serializable props include:
> ...
> - Client or Server Component elements (JSX)

So yes, you can return components from Server Actions

**Reply by rohitpotato** [reply]
Thanks! My concern here is, is it a correct approach to do at scale? Moving everything to server means increasing server througput, is the optimisation worth enough to implement at scale?

**Reply by Juneezee** [reply]
> is it a correct approach to do at scale?

Well, the React docs https://react.dev/reference/rsc/use-server#use-server says:

> Server Actions are designed for mutations that update server-side state; they are not recommended for data fetching. Accordingly, frameworks implementing Server Actions typically process one action at a time and do not have a way to cache the return value.

I wouldn't say it is incorrect. I do think Server Actions provide a better developer experience than API Routes. I couldn't answer regarding the optimisation part because I haven't done / seen any large scale benchmark yet.

I couldn't find the part on official Next.js docs where it mention that it processes Server Actions one at a time. There is this blog though: https://tigerabrodi.blog/nuances-of-server-actions-in-nextjs#heading-works-like-a-queue

**Comment by rohitpotato** [comment]
@leerob @samcx Can you weight in on this? Would love to hear your take on this.

**Reply by samcx** [reply]
@rohitpotato

> Server Actions are designed for mutations that update server-side state; they are not recommended for data fetching.

I don't see an issue with this statement as well. Ideally, for simplicity, your Server Action would mutate your server data, then your `fetch()` in your Component fetch the latest data again. In the case of the AI SDK, it's used to stream data off one Server Action—it's not intended to fire off multiple Actions in a quick fashion.

Returning a Component also is not a "bad practice"—if it's serializable, I don't see why we couldn't return a Component. 

> the fact that data-fetching is not recommended + they are executed in a queue can cause performance issues and bottleneck at the infra end. Our product receives very high traffic, and core web vitals are non-negotiable

Returning a Server Component ≠ fetching data in your Server Action, so this is a correct take in my opinion.

> And also, no caching. Does that mean that the browser wont be able to cache the resources?

The Component responsible for receiving the Component from the Server Action should be responsible for caching.

> Our main selling point was server to server communication, data-fetching at the server, generate the rsc payload and send it over the network. This should reduce client overhead in low connectivity areas.

Wouldn't just fetching JSON data in a `fetch()` be less data than fetching additional RSCs? Fetch the Client data (the Component) once, and re-render with new JSON data as needed. I guess it really depends. You can also look to optimize fetching the data (the backend endpoint).

**Reply by rohitpotato** [reply]
@samcx So the usecase was to lazy load different rsc components, with each pagination requests, the attempt is to make a call to the server action, fetch the data inside the server actions, pass the data to the appropriate server component (widget), and return it. 

If it were the same component rendered with pagination, it would have better to just return the data or even do client side pagination. 

Also, is there a overhead on rendering js vs reconciling rsc for low end devices?

**Reply by samcx** [reply]
> Also, is there a overhead on rendering js vs reconciling rsc for low end devices?

I doubt there's any significance overhead to downloading and analyzing the React Server Component Payload (RSC Payload). Have not heard of any report like this. The RSC Payload is a compact binary representation of the rendered React Server Components tree.

**Comment by icyJoseph** [comment]
Interesting, 

- https://react.dev/reference/rsc/use-server#noun-labs-1201738-(2)

> Server Actions are designed for mutations that update server-side state; they are not recommended for data fetching. Accordingly, frameworks implementing Server Actions typically process one action at a time and do not have a way to cache the return value.

I guess, one could just return an object, with the data, for the next thing, and go from there. Using a switch case to load the lazy loaded components. A React element is just a plain object representation of your UI element, and carrying that over the network, I guess, kind of works 😅 but RSC was made for that very purpose. RSC is stringified VDOM.

**Reply by rohitpotato** [reply]
the fact that data-fetching is not recommended + they are executed in a queue can cause performance issues and bottleneck at the infra end. Our product receives very high traffic, and core web vitals are non-negotiable. 

And also, no caching. Does that mean that the browser wont be able to cache the resources? 

Our main selling point was server to server communication, data-fetching at the server, generate the rsc payload and send it over the network. This should reduce client overhead in low connectivity areas.

But at this point, the trade-offs between client vs server load is not worth the investment.

If in future, they do decide to somehow optimise this, say by making this a first-class thing, this can be a really handy feature.

Also, the vercel ai-sdk has a streaming util, which can stream rsc payload over the n/w using server actions. But that unfortunately is tied in with the ai-sdk.

This is what brought us to exploring this in the first place.
